services/autoscaling/src/simcore_service_autoscaling/modules/dask.py

```diff
@@ -118,19 +81,87 @@ def _find_by_worker_host(
     ) -> bool:
         _, details = dask_worker
-        return bool(details["host"] == node_ip) or bool(
-            node_hostname in details["name"]
-        )
+        if match := re.match(DASK_NAME_PATTERN, details["name"]):
+            return bool(match.group("private_ip") == node_hostname)
+        return False
 
     filtered_workers = dict(filter(_find_by_worker_host, workers.items()))
     if not filtered_workers:
         raise DaskWorkerNotFoundError(
             worker_host=ec2_instance.aws_private_dns, url=client.scheduler.address
         )
-    assert len(filtered_workers) == 1  # nosec
+    assert (
+        len(filtered_workers) == 1
+    ), f"returned workers {filtered_workers}, {node_hostname=}"  # nosec
     return next(iter(filtered_workers.items()))
 
 
+async def is_worker_connected(
+    scheduler_url: AnyUrl, worker_ec2_instance: EC2InstanceData
+) -> bool:
+    with contextlib.suppress(DaskNoWorkersError, DaskWorkerNotFoundError):
+        async with _scheduler_client(scheduler_url) as client:
+            _dask_worker_from_ec2_instance(client, worker_ec2_instance)
+            return True
+    return False
+
+
+async def list_unrunnable_tasks(url: AnyUrl) -> list[DaskTask]:
+    """
+    Raises:
+        DaskSchedulerNotFoundError
+    """
+
+    def _list_tasks(
+        dask_scheduler: distributed.Scheduler,
+    ) -> dict[str, dict[str, Any]]:
+        return {
+            task.key: task.resource_restrictions for task in dask_scheduler.unrunnable
+        }
+
+    async with _scheduler_client(url) as client:
+        list_of_tasks: dict[
+            DaskTaskId, DaskTaskResources
+        ] = await _wrap_client_async_routine(client.run_on_scheduler(_list_tasks))
+        _logger.debug("found unrunnable tasks: %s", list_of_tasks)
+        return [
+            DaskTask(task_id=task_id, required_resources=task_resources)
+            for task_id, task_resources in list_of_tasks.items()
+        ]
+
+
+async def list_processing_tasks_per_worker(
+    url: AnyUrl,
+) -> dict[DaskWorkerUrl, list[DaskTask]]:
+    """
+    Raises:
+        DaskSchedulerNotFoundError
+    """
+
+    def _list_tasks(
+        dask_scheduler: distributed.Scheduler,
+    ) -> dict[str, list[tuple[DaskTaskId, DaskTaskResources]]]:
+        worker_to_processing_tasks = defaultdict(list)
+        for task_key, task_state in dask_scheduler.tasks.items():
+            if task_state.processing_on:
+                worker_to_processing_tasks[task_state.processing_on.address].append(
+                    (task_key, task_state.resource_restrictions)
+                )
+        return worker_to_processing_tasks
+
+    async with _scheduler_client(url) as client:
+        worker_to_tasks: dict[
+            str, list[tuple[DaskTaskId, DaskTaskResources]]
+        ] = await _wrap_client_async_routine(client.run_on_scheduler(_list_tasks))
+        _logger.debug("found processing tasks: %s", worker_to_tasks)
+        tasks_per_worker = {}
```

Comment:
MINOR: you can use intronspection here for tasks_per_worker

Reply:
not sure I follow whatyou mean here